api/saml/configuration/model.py

```diff
@@ -0,0 +1,611 @@
+        description=_(
+            "If strict is 1, then the Python Toolkit will reject unsigned or unencrypted messages "
+            "if it expects them to be signed or encrypted. Also, it will reject the messages "
+            "if the SAML standard is not strictly followed."
+        ),
+        type=ConfigurationAttributeType.NUMBER,
+        required=False,
+        default=0,
+    )
+
+    service_provider_debug_mode = ConfigurationMetadata(
+        key="debug",
+        label=_("Service Provider's Debug Mode"),
+        description=_("Enable debug mode (outputs errors)."),
+        type=ConfigurationAttributeType.NUMBER,
+        required=False,
+        default=0,
+    )
+
+    IDP_DISPLAY_NAME_DEFAULT_TEMPLATE = "Identity Provider #{0}"
+
+    def __init__(self, configuration_storage, db, metadata_parser):
+        """Initializes a new instance of SAMLConfiguration class
+
+        :param configuration_storage: SAML configuration storage
+        :type configuration_storage: ConfigurationStorage
+
+        :param metadata_parser: SAML metadata parser
+        :type metadata_parser: SAMLMetadataParser
+        """
+        super(SAMLConfiguration, self).__init__(configuration_storage, db)
+
+        self._metadata_parser = metadata_parser
+
+        self._identity_providers = None
+        self._service_provider = None
+
+    def _get_federated_identity_providers(self, db):
+        """Return a list of federated IdPs corresponding to the entity IDs selected by the admin.
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: List of SAMLFederatedIdP objects
+        :rtype: List[api.saml.metadata.federations.model.SAMLFederatedIdP]
+        """
+        if not self.federated_identity_provider_entity_ids:
+            return []
+
+        federated_identity_provider_entity_ids = json.loads(
+            self.federated_identity_provider_entity_ids
+        )
+
+        return (
+            db.query(SAMLFederatedIdentityProvider)
+            .filter(
+                SAMLFederatedIdentityProvider.entity_id.in_(
+                    federated_identity_provider_entity_ids
+                )
+            )
+            .all()
+        )
+
+    def _load_identity_providers(self, db):
+        """Loads IdP settings from the library's configuration settings
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: List of IdentityProviderMetadata objects
+        :rtype: List[IdentityProviderMetadata]
+
+        :raise: SAMLParsingError
+        """
+        identity_providers = []
+
+        if self.non_federated_identity_provider_xml_metadata:
+            parsing_results = self._metadata_parser.parse(
+                self.non_federated_identity_provider_xml_metadata
+            )
+            identity_providers = [
+                parsing_result.provider for parsing_result in parsing_results
+            ]
```

Comment:
Note to myself: need to make sure that this is only parsed during LibraryAuthenticator construction.